### ISSUE TITLE
I would love support for Cost Sensitive Classification

### ISSUE THREAD

**Issue by simonhughes22** [amzn/pecos#34]
## Description
Reading through the code it looks like the base XR linear models support assigning a cost to each data point so that they can be weighted differently, e.g. by miss classification cost. In a search and ranking setting, some labels may be more important than others, e,g. applies are more important than clicks for a job board, or movies watched are more important than movies clicked on for a streaming service. However, the more important class is often much sparser, and so being able to use both a positive instances, one as strong one as weak positives, would improve the utility of the model.

Right now, all training instances are treated equally, but if a weight matrix could be provided with instance weights, or a labelling scheme were training data points have different labels, and each label has a relative weight, that would meet this need. The code references a R matrix as input to MLmodel.train, but it cannot be passed in from the trainer.

Comments in MLProblem.init:

```python
Args:
            X (csr_matrix, np.ndarray or ScipyDrmF32): Instance feature matrix.
            Y (csr_matrix, np.ndarray or ScipyCscF32): Instance-to-label matrix.
            C (csc_matrix, np.ndarray or ScipyCscF32, optional): Label-to-cluster matrix.
                If not given, create an all-one matrix of shape `(Y.shape[1], 1)`.
            M (csc_matrix, np.ndarray or ScipyCscF32, optional): Instance-to-cluster matrix.
                If not given, creates M from Y*C with multi-threading sparse_matmul.
            R (csc_matrix, np.ndarray or ScipyCscF32, optional): Relevance matrix.
                If not given, will use None.
            threads(int, optional): Number of threads for multi-threading. Default to 8.
```
I would like to be able to pass in R. However in HierarchicalMLModel.train (line 1186 in xmc/base.py) the following code prevents this (and it's not exposed in calling classes):
```
        if prob.R is not None:
            raise NotImplementedError(
                "Cost-senstive learning for HierarchicalMLModel is not yet supported"
            )
```

Is this just a case of exposing the variable to the different train classes in the parents? If so i would happily implement that myself (and issues a PR) but i was thinking there was probably a reason for it to be hidden. If it's that simple please let me know here and i will work on adding it in.

**Comment by OctoberChang** [comment]
@simonhughes22 , thanks for your interest in PECOS and the usage of cost-sensitive learning!
As of `version=v0.1.0`, we did not support cost-sensitive learning at the high level Python API `HierarchicalMLModel.train` because of three reasons:
(1) How to design a proper relevance matrix `R` is very task-specific; a  poorly-designed `R` may deteriorate performance.
(2) The scale of relevance matrix `R` will affect balance between the loss function and the regularization term. Thus,  `Cp`  in PECOS needed to be tuned carefully.
(3) The value of relevance matrix needed to be greater than zero! Otherwise the loss function is no longer convex.

With those notes in mind, here is an example code snippet of cost-sensitive learning on the `wiki10-31k` XMC dataset.
``` python
import copy
import numpy as np
from sklearn.preprocessing import normalize
from pecos.core import clib
from pecos.utils import smat_util
from pecos.xmc import Indexer, LabelEmbeddingFactory
from pecos.xmc import HierarchicalMLModel, MLProblem, MLModel

# load your data
X_trn = smat_util.load_matrix("./wiki10-31k/attnxml/X.trn.npz")
X_tst = smat_util.load_matrix("./wiki10-31k/attnxml/X.tst.npz")
Y_trn = smat_util.load_matrix("./wiki10-31k/Y.trn.npz")
Y_tst = smat_util.load_matrix("./wiki10-31k/Y.tst.npz")
R_trn = copy.deepcopy(Y_trn) # only for XMC benchmarks, need to define your own!

# get cluster chain
label_feat = LabelEmbeddingFactory.create(Y_trn, X_trn, method="pifa")
C_chain = Indexer.gen(label_feat, "hierarchicalkmeans", nr_splits=32, max_leaf_size=100)

# one way to get relevance matrix chain for the hierarchical label tree of XR-Linear models
Y_chain, R_chain = [Y_trn], [R_trn]
for C in reversed(C_chain[1:]):
  Y_t = clib.sparse_matmul(Y_chain[-1], C, threads=32)
  Y_chain.append(Y_t.tocsc())
  # this is a heuristic we found useful for XMC benchmarks
  R_t = clib.sparse_matmul(R_chain[-1], C, threads=32)
  R_chain.append(R_t)
Y_chain.reverse()
R_chain = [normalize(R_t, norm="l1", axis=1) for R_t in R_chain[::-1]]

DO_COST_SENSITIVE = True
# cost-sensitive learning needs to tune the Cp hyper-parameters
Cp_list = [1.0, 8.0, 4.0] if DO_COST_SENSITIVE else [1.0, 1.0, 1.0]
model_chain = []
for t, (Y_t, C_t, R_t) in enumerate(zip(Y_chain, C_chain, R_chain)):
    train_params = MLModel.TrainParams.from_dict({"Cp": Cp_list[t]})
    pred_params = MLModel.PredParams()
    R_t = R_t if DO_COST_SENSITIVE else None 
    if t == 0:
        cur_prob = MLProblem(X_trn, Y_t, R=R_t)
    else:
        M_t = smat_util.binarized(Y_chain[t - 1].tocsc())
        cur_prob = MLProblem(X_trn, Y_t, R=R_t, C=C_t, M=M_t)
    cur_model = MLModel.train(cur_prob, train_params, pred_params)
    model_chain.append(cur_model)

hlm = HierarchicalMLModel(model_chain)
Y_tst_pred = hlm.predict(X_tst)
print(smat_util.Metrics.generate(Y_tst, Y_tst_pred))
```
Here are the evaluation results:
```bash
# for cost-sensitive learning, we got
# prec  = 85.72 80.86 75.01 70.03 65.24 61.45 57.90 54.67 51.83 49.27
# recall = 5.08 9.52 13.10 16.18 18.69 20.99 22.95 24.66 26.18 27.56

# For original xrlinear, the numbers are
# prec  = 84.54 78.98 73.13 68.27 64.10 60.53 57.22 54.26 51.65 49.21
# recall = 5.00 9.28 12.75 15.73 18.35 20.66 22.65 24.44 26.05 27.48
```

We are still investigating if there's a stable/robust default hyper-parameters for the cost-senstive learning usage.
And after the investigation, we may consider implementing this functionality at the higher level Python API such as `HierarchicalMLModel.train`. 

If you find any successful case of using cost-sensitive PECOS on your applications, please let us know.
We would be very happy to hear your experience!

**Comment by simonhughes22** [comment]
Awesome thanks @OctoberChang.
